### PR TITLE
[RELENG-7490] ✨ Workflow cost

### PR DESCRIPTION
**Pull request by gaspardmoindrot**

### CODE REVIEW

File: gh_actions_exporter/config.py [hunk 1]
```diff
@@ -45,6 +45,12 @@ class Settings(BaseSettings):
     }
     flavor_label: Optional[str] = 'flavor'
     default_cost: Optional[float] = 0.008
+    github_app_private_pem: SecretStr # Checker
+    github_app_id: int
+    github_app_installation_id: int
+    title: str = "Workflow Costs"
```
Comment (tcarmet) [inline gh_actions_exporter/config.py:51]: Both `title` and `summary` variables should be giving more context than just title as there might be multiple check-runs in the future with other type of stats, not cost related.

File: gh_actions_exporter/cost.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                labels.update(self.relabel_job_labels(relabel, job_request["labels"]))
+            elif relabel.type == RelabelType.name:
+                labels.update(self.relabel_job_names_dict(relabel, job_request))
+        return labels
+
+    def _get_job_cost(self, job_request: dict, webhook: WebHook) -> float:
+        labels: dict = self.job_request_labels(job_request, webhook, self.settings)
+        flavor = labels.get(self.settings.flavor_label, None)
+        cost_per_min: float = self.settings.job_costs.get(flavor, self.settings.default_cost)
+        duration: float = (job_request["completed_at"].timestamp()
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:57]: job request is not typed. It seems like a [WorkflowJob](https://github.com/scality/gh-actions-exporter/pull/20/files/4dde2b798b1473e1b66bf462f1faf1d144931549..3e7762f1f426b80668bb0f617366c74606c98c69#diff-492b94ab9c36f44dc4552990feb27270cf5cf3f035de199e212362e0260cc9c9R16) type of object. If so, use it.

File: gh_actions_exporter/cost.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        url: str = f"/repos/{webhook.repository.full_name}/commits/{webhook.workflow_run.head_sha}/check-runs"
+        check_runs: dict = repository._requester.requestJsonAndCheck("GET", url, {"Accept": "application/vnd.github.v3+json",})[1]
+
+        for check_run in check_runs["check_runs"]:
+            if check_run["name"] == "self.settings.title" and check_run["app"]["id"] == self.settings.github_app_id:
+                return check_run["output"]["summary"]
+        return ""
+
+    def upload_check_run(self, summary: str, webhook: WebHook, access_token: str) -> None:
+        g: Github = Github(login_or_token=access_token)
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:72]: GitHub class should be instanciated as a fastapi dependency. https://fastapi.tiangolo.com/tutorial/dependencies/

File: gh_actions_exporter/cost.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                "summary": summary
+            }
+        )
+
+    def display_cost(self, webhook: WebHook):
+        if webhook.workflow_run.conclusion:
+            token_generator: TokenGenerator = TokenGenerator(self.settings)
+            access_token: str = token_generator.generate_token()
+            
+            g: Github = Github(access_token)
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:91]: Authentication to GitHub should be done with the appropriate method provided by the class. 

Using the `AppAuthentication` class. Example:
```python
  Github(
    app_auth=AppAuthentication(
        app_id=self.settings.github_app_id,
        private_key=self.settings.github_private_pem,
        installation_id=github_app_installation_id,
    ),
)
```
https://github.com/PyGithub/PyGithub/blob/300c5015d41236eb410b9cbe895258294e3500af/tests/Authentication.py#L55

File: gh_actions_exporter/config.py [hunk 5]
```diff
@@ -45,6 +45,12 @@ class Settings(BaseSettings):
     }
     flavor_label: Optional[str] = 'flavor'
     default_cost: Optional[float] = 0.008
+    github_app_private_pem: SecretStr # Checker
```
Comment (tcarmet) [inline gh_actions_exporter/config.py:48]: for naming consistency with your library
```suggestion
    github_app_private_key: SecretStr # Checker
```

File: gh_actions_exporter/cost.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
+import jinja2
+
+from typing import Dict, List
+from gh_actions_exporter.config import Relabel, RelabelType, Settings
+from gh_actions_exporter.types import WebHook, JobCost
+from gh_actions_exporter.tokenGenerator import TokenGenerator
+from github import Github, Repository
+
+class Cost(object):
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:9]: while having a cost class is a good idea, not all functions are related to cost. Some are merely relabelling functions.

File: gh_actions_exporter/cost.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            summary: str = self.get_previous_check_run(repository, webhook)
+
+            template = jinja2.Template(
+                """{% if summary == "" %}
+                {{ settings.summary }}
+                {% endif %}
+                ## [{{ workflow_run.name }}]({{ workflow_run.html_url }})
+                {% for job_cost in jobs_cost %}* {{ job_cost.name }}: ${{ job_cost.cost }}
+                {% endfor %}Total: ${{ total_cost }}"""
+            )
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:112]: Configure the content of the template as a `workflow_cost.md.j2` file.

File: gh_actions_exporter/cost.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+            data: dict = {
+                "summary": summary,
+                "settings": {"summary": self.settings.summary},
+                "workflow_run": {"name": webhook.workflow_run.name, "html_url": webhook.workflow_run.html_url},
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:117]: why not?

I understand you only want to mention what you need, but I see no harm in keeping the whole workflow run object available if needed.
```suggestion
                "workflow_run": webhook.workflow_run,
```

File: gh_actions_exporter/cost.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            template = jinja2.Template(
+                """{% if summary == "" %}
+                {{ settings.summary }}
+                {% endif %}
+                ## [{{ workflow_run.name }}]({{ workflow_run.html_url }})
+                {% for job_cost in jobs_cost %}* {{ job_cost.name }}: ${{ job_cost.cost }}
+                {% endfor %}Total: ${{ total_cost }}"""
+            )
+
+            data: dict = {
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:114]: Convert data into a pydantic basemodel

**Review by tcarmet — CHANGES_REQUESTED** [review]

File: gh_actions_exporter/cost.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        flavor = labels.get(self.settings.flavor_label, None)
+        cost_per_min: float = self.settings.job_costs.get(flavor, self.settings.default_cost)
+        duration: float = (job_request["completed_at"].timestamp()
+                    - job_request["started_at"].timestamp())
+
+        return duration / 60 * cost_per_min
+
+    def get_previous_check_run(self, repository: Repository, webhook: WebHook) -> str:
+        url: str = f"/repos/{webhook.repository.full_name}/commits/{webhook.workflow_run.head_sha}/check-runs"
+        check_runs: dict = repository._requester.requestJsonAndCheck("GET", url, {"Accept": "application/vnd.github.v3+json",})[1]
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:64]: Use the appropriate method: https://pygithub.readthedocs.io/en/latest/github_objects/CheckRun.html#checkrun

File: gh_actions_exporter/config.py (hunk 1 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/config.py:62]: Exactly, there will be other `check-runs`, so in other words, the `title` and `summary` here will have nothing to do with the others

File: gh_actions_exporter/cost.py (hunk 6 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/cost.py:9]: I think now it's better, we only have `_get_job_cost` & `_runner_type_job` not related this much

Comment (gaspardmoindrot) [reply gh_actions_exporter/cost.py:9]: Tell me if it's ok for you now, otherwise tell me what you want for me to do

File: tests/api/test_cost.py
Comment (tcarmet) [inline tests/api/test_cost.py:1]: I see no API call so feel free to rename this file into `tests/unit/test_cost.py`

File: gh_actions_exporter/config.py [hunk 11]
```diff
@@ -35,6 +38,8 @@ class ConfigFile(BaseSettings):
 
 
 class Settings(BaseSettings):
+    load_dotenv()
```
Comment (tcarmet) [inline gh_actions_exporter/config.py:41]: For support of `.env` files use the method documented by the libray: https://docs.pydantic.dev/latest/usage/settings/#dotenv-env-support

**Review by tcarmet — CHANGES_REQUESTED** [review]

File: gh_actions_exporter/cost.py (hunk 4 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/cost.py]: Done

File: gh_actions_exporter/config.py (hunk 11 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/config.py:41]: Ok, I'm working on it !

Comment (gaspardmoindrot) [reply gh_actions_exporter/config.py:41]: Done

File: gh_actions_exporter/cost.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        duration: float = (datetime.fromisoformat(job_request["completed_at"].replace('Z', '+00:00')).timestamp()
+                    - datetime.fromisoformat(job_request["started_at"].replace('Z', '+00:00')).timestamp())
+
+        if self._runner_type_job(job_request) == "github-hosted":
+            return duration / 60 * self.settings.default_cost
+        return duration / 60 * cost_per_min
+
+    def _get_workflow_jobs(self, webhook: WebHook, g: GitHub) -> dict:
+        workflow_jobs: dict = g.rest.actions.list_jobs_for_workflow_run(webhook.organization.login, webhook.repository.name, webhook.workflow_run.id).json()
+        return workflow_jobs
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py]: Not typed.
This function should return: `List[WorkflowJob]`

File: gh_actions_exporter/types.py [hunk 13]
```diff
@@ -54,8 +55,24 @@ class Repository(BaseModel):
...
+class JobCost(BaseModel):
+    name: str
+    cost: int
+
+class CheckRunData(BaseModel):
+    summary: str
+    settings: dict
+    workflow_run: dict
+    jobs_cost: list[dict]
+    total_cost: float
```
Comment (tcarmet) [inline gh_actions_exporter/types.py:80]: All fields set as dict are not typed enough, what's inside the dict?

File: gh_actions_exporter/main.py [hunk 14]
```diff
@@ -35,13 +41,15 @@ async def webhook(
     webhook: WebHook,
     request: Request,
     settings: Settings = Depends(get_settings),
-    metrics: Metrics = Depends(metrics)
+    metrics: Metrics = Depends(metrics),
+    cost: Cost = Depends(cost),
```
Comment (tcarmet) [inline gh_actions_exporter/main.py:45]: I don't understand why cost is set as fastapi dependency. Because it feels to me that it can be called like a normal function from anywhere. (I could be wrong though).

File: gh_actions_exporter/tokenGenerator.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+    def generate_token(self) -> GitHub:
+        github = GitHub(
+            AppInstallationAuthStrategy(
+                self.github_app_id,
+                self.github_app_private_key,
+                self.github_app_installation_id,
+            )
+        )
+        return github
```
Comment (tcarmet) [inline gh_actions_exporter/tokenGenerator.py:23]: * This whole class should actually be a function defined as a fastapi dependency. Similar to what you did with cost.
* The name does not fit its function. You talk about tokengenrator but you return a client.

File: gh_actions_exporter/types.py
Comment (tcarmet) [inline gh_actions_exporter/types.py:1]: A lot of the classes defined in this file, WebHook, WorkflowRun, etc etc, can be replaced by models defined by the githubkit library. As the same models are available inside that class.

File: workflow_cost.md.j2
Comment (tcarmet) [inline workflow_cost.md.j2:1]: move this file to `templates/workflow_cost.md.j2`

**Review by tcarmet — CHANGES_REQUESTED** [review]
Did a quick run over the pr.

The logic is looking good, the work remains on knowledge of certain tools or library that needs to be applied as well as ensuring that typing is used accordingly.

File: gh_actions_exporter/cost.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+                "title": self.settings.title,
+                "summary": summary
+            }
+        }
+        g.rest.checks.create(webhook.organization.login, webhook.repository.name, data=data)
+
+    def display_cost(self, webhook: WebHook):
+        if webhook.workflow_run.conclusion:
+            token_generator: TokenGenerator = TokenGenerator(self.settings)
+            g: GitHub = token_generator.generate_token()
```
[the remark this answers is not in the document]
Comment (tcarmet) [reply gh_actions_exporter/cost.py:89]: Actually you are, it was hidden behind the function. Wrote a new comment with the naming and usage of token_generator

File: tests/unit/test_cost.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+        "head_sha": "1",
+        "url": "1",
+        "html_url": "1",
+        "status": "1",
+        "started_at": "2021-11-16T17:51:47Z",
+        "completed_at": "2021-11-16T17:59:47Z",
+        "name": "1",
+        "labels": [
+            "ubuntu-latest",
+            "self-hosted"
```
[the remark this answers is not in the document]
Comment (gaspardmoindrot) [reply tests/unit/test_cost.py]: Oh yeah, you're right, I change for something better

File: gh_actions_exporter/types.py (hunk 13 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/types.py:80]: Oh ok, I thought putting just `dict` was enough, I'll change for every occurence I can find

File: gh_actions_exporter/main.py (hunk 14 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/main.py:45]: You're right, it's fixed

File: gh_actions_exporter/tokenGenerator.py (hunk 15 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/tokenGenerator.py:23]: Done! Tell me if it's better now

File: workflow_cost.md.j2
Comment (gaspardmoindrot) [reply workflow_cost.md.j2:1]: Done!

File: gh_actions_exporter/cost.py [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+            flavor_label, self.settings.default_cost
+        )
+        duration: float = (
+            datetime.fromisoformat(
+                job_request["completed_at"].replace("Z", "+00:00")
+            ).timestamp()
+            - datetime.fromisoformat(
+                job_request["started_at"].replace("Z", "+00:00")
+            ).timestamp()
+        )
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:34]: not typed.

Comment (tcarmet) [reply gh_actions_exporter/cost.py:34]: example: job_request.completed_at.replace("Z", "+00:00")

File: gh_actions_exporter/cost.py [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+        return ""
+
+    def _upload_check_run(self, summary: str, g: GitHub, webhook: WebHook) -> None:
+        data = {
+            "name": "Cost",
+            "head_sha": webhook.workflow_run.head_sha,
+            "status": "completed",
+            "conclusion": "success",
+            "output": {"title": self.settings.title, "summary": summary},
+        }
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:83]: type it, and you can create a new basemodel class if needed.

File: gh_actions_exporter/tokenGenerator.py (hunk 15 above)
Comment (tcarmet) [reply gh_actions_exporter/tokenGenerator.py]: Are we missing a git push?

File: tests/api/metrics/conftest.py [hunk 20]
```diff
@@ -93,10 +80,11 @@ def workflow_run():
             "updated_at": "2021-11-16T17:53:32Z",
         },
         "repository": {
-            "name": "repo",
-            "full_name": "org/repo",
+            "name": "runners-test",
+            "full_name": "scalanga-devl/runners-test",
```
Comment (tcarmet) [inline tests/api/metrics/conftest.py:84]: Anonymize the tests cases, revert the changes

File: tests/api/metrics/test_app.py [hunk 21]
```diff
@@ -0,0 +1,25 @@
+def test_unsupported_events(client):
```
Comment (tcarmet) [inline tests/api/metrics/test_app.py:1]: as those tests are not necessarily related to metrics, this particular file can be moved to something around the lines of `tests/api/app/test_app.py`

File: gh_actions_exporter/main.py [hunk 22]
```diff
@@ -19,10 +21,15 @@ def metrics() -> Metrics:
     return Metrics(get_settings())
 
 
+@lru_cache()
+def github_client() -> GithubClient:
+    return GithubClient(get_settings())
+
+
```
Comment (tcarmet) [inline gh_actions_exporter/main.py:28]:
```suggestion
@lru_cache()
def github() -> GitHub:
    # insert logic contained inside the GitHubClient class
    


```

**Review by tcarmet — CHANGES_REQUESTED** [review]

File: gh_actions_exporter/cost.py [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+        g.rest.checks.create(
+            webhook.organization.login, webhook.repository.name, data=data
+        )
+
+    def display_cost(self, webhook: WebHook, github_client: GithubClient):
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:88]:
```suggestion
    def display_cost(self, webhook: WebHook, github: GithubClient):
```

File: gh_actions_exporter/cost.py [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+        if webhook.workflow_run.conclusion:
+            #token_generator: GithubClient = GithubClient(self.settings)
+            g: GitHub = github_client.get_client()
```
Comment (tcarmet) [inline gh_actions_exporter/cost.py:91]: All reference to the github client should be named as `github`

**Review by tcarmet — CHANGES_REQUESTED** [review]

File: gh_actions_exporter/cost.py (hunk 19 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/cost.py:83]: Done

File: tests/api/metrics/test_app.py (hunk 21 above)
Comment (gaspardmoindrot) [reply tests/api/metrics/test_app.py:1]: Done

File: gh_actions_exporter/main.py (hunk 22 above)
Comment (gaspardmoindrot) [reply gh_actions_exporter/main.py:28]: Better now

**Review by tcarmet — APPROVED** [review]
